lib/main.js

```diff
@@ -0,0 +1,152 @@
+    this.subscriptions.add(atom.config.observe('linter-jshint.lintJSXFiles', lintJSXFiles => {
+      this.lintJSXFiles = lintJSXFiles
+      if (lintJSXFiles) {
+        this.scopes.push(scopeJSX)
+      } else {
+        if (this.scopes.indexOf(scopeJSX) !== -1) {
+          this.scopes.splice(this.scopes.indexOf(scopeJSX), 1)
+        }
+      }
+    }))
+
+    const scopeEmbedded = 'source.js.embedded.html'
+    this.subscriptions.add(
+      atom.config.observe('linter-jshint.lintJSXFiles', lintInlineJavaScript => {
+        this.lintInlineJavaScript = lintInlineJavaScript
+        if (lintInlineJavaScript) {
+          this.scopes.push(scopeEmbedded)
+        } else {
+          if (this.scopes.indexOf(scopeEmbedded) !== -1) {
+            this.scopes.splice(this.scopes.indexOf(scopeEmbedded), 1)
+          }
+        }
+      })
+    )
+  },
+  deactivate() {
+    this.subscriptions.dispose()
+  },
+  provideLinter(): Linter$Provider {
+    const Helpers = require('atom-linter')
+    const Reporter = require('jshint-json')
+
+    return {
+      name: 'JSHint',
+      grammarScopes: this.scopes,
+      scope: 'file',
+      lintOnFly: true,
+      lint: async (textEditor) => {
+        const results = []
+        const filePath = textEditor.getPath()
+        const fileContents = textEditor.getText()
+        const parameters = ['--reporter', Reporter, '--filename', filePath]
+
+        let configFile = null
+        if (this.disableWhenNoJshintrcFileInPath) {
+          configFile = await Helpers.findCachedAsync(Path.dirname(filePath), '.jshintrc')
+          if (configFile) {
+            parameters.push('--config', configFile)
+          } else {
+            return results
+          }
+        }
+        if (this.lintInlineJavaScript) {
+          parameters.push('--extract', 'always')
+        }
+        parameters.push('-')
+
+        const result = await Helpers.execNode(
+          this.executablePath, parameters, { stdin: fileContents }
+        )
+        let parsed
+        try {
+          parsed = JSON.parse(result)
+        } catch (_) {
+          console.error('[Linter-JSHint]', _, result)
+          atom.notifications.addWarning('[Linter-JSHint]',
+            { detail: 'JSHint return an invalid response, check your console for more info' }
+          )
+          return results
+        }
+
+        for (const entry of parsed.result) {
+          if (!entry.error.id) {
+            continue
+          }
+
+          const error = entry.error
+          const errorLine = error.line - 1
+          let range
+
+          // TODO: Remove workaround of jshint/jshint#2846
+          if (error.character === null) {
+            range = Helpers.rangeFromLineNumber(textEditor, errorLine)
+          } else {
+            let character = error.character
+            const maxCharacter = textEditor.getBuffer().lineLengthForRow(errorLine)
+            // TODO: Remove workaround of jshint/jshint#1457
```

Comment:
Also, this should be above the `maxCharacter` determination, as that is part of the workaround.